### PR TITLE
Updating of artifacts and artifact data

### PR DESCRIPTION
**Pull request by MorpheusXAUT**
# TL;DR
This PR extends datacatalog to support the updating of artifacts and overwriting of stored artifact data as well as some minor cleanup/housekeeping.

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [X] Any pending items have an associated Issue

## Complete description
The new `UpdateArtifact` endpoints allows for an existing artifact to be updated. At the moment, only the underlying `ArtifactData` can be overwritten.
While updating, datacatalog will create any new data keys, overwrite any existing ones with the new data provided and remove no longer provided elements from the underlying blob storage. To facilitate a full cleanup, deletion of artifact data was implemented as well.

This PR also fixes the mock generation for the repository interfaces as they seem to have never been re-generated using the latest `mockery` version.
Additionally, the datacatalog project was moved to Go 1.18 and the Dockerfile setup was adapted to be in line with flyteadmin. Some minor cleanup/housekeeping was performed to make the linter happy.

As this uses a currently unmerged [`flyteidl`](https://github.com/flyteorg/flyteidl/pull/318) version, the PR is created as a draft (for review) until a new version of `flyteidl` is available.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2867

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/manager/impl/artifact_manager.go [hunk 1]
```diff
@@ -283,6 +308,97 @@ func (m *artifactManager) ListArtifacts(ctx context.Context, request *datacatalo
 	return &datacatalog.ListArtifactsResponse{Artifacts: artifactsList, NextToken: token}, nil
 }
 
+// UpdateArtifact updates the given artifact, currently only allowing the associated ArtifactData to be replaced. All
+// stored data will be overwritten in the underlying blob storage, no longer existing data (based on ArtifactData name)
+// will be deleted.
+func (m *artifactManager) UpdateArtifact(ctx context.Context, request *datacatalog.UpdateArtifactRequest) (*datacatalog.UpdateArtifactResponse, error) {
```
Comment (hamersaw) [inline pkg/manager/impl/artifact_manager.go:314]: Single comment on this function - IIUC we are updating by:
1. storing new ArtifactData (while updating the inmemory Artifact model)
2. deleting the existing existing data
3. updating the db Artifact
I think this could cause a corrupt system if there is a failure between 2 and 3. If we delete the existing ArtifactData and then fail to update the model then the next cache lookup will look for non-existent data. Does it make sense to switch steps 2 and 3? Then the biggest issue would be orphaned data in case of a failure - which could be easily cleaned up by searching the ArtifactData for an existing link.

Comment (MorpheusXAUT) [reply pkg/manager/impl/artifact_manager.go:314]: Basically, yeah.

We are:
1. Creating a "diff" between provided data and already existing one (based on artifact data name)
2. Creating "new" artifact data and overwriting existing one (in blob storage)
3. Deleting no longer existing artifact data (from blob storage)
4. Updating the database models (artifact and artifact data) to reflect the changes above

You're right about the broken state in case of an error, unfortunately we can't run in a transaction using the data store :sweat_smile:. Switching the order would work, I guess, however we might also end up with partially updated data (some of the new data was written successfully while parts of it haven't been updated yet) or artifact with both new _and_ old data (including results that should've been deleted already).

I guess that's still the best option for a partial failure like this (aside from completely nuking all cached data when encountering an error, which could potentially fail as well though).

Comment (hamersaw) [reply pkg/manager/impl/artifact_manager.go:314]: > unfortunately we can't run in a transaction using the data store
Could not agree more! I noticed the `CreateArtifact` implementation can have this same issue (orphaned data on failure). I don't think this is common issue now, but if we find it is we will have to implement some kind of garbage collection mechanism.

> I guess that's still the best option for a partial failure like this
Yeah, If you have other thoughts I'm very open. This is the best approach that I could think of - I think a partial update is preferential to a cache hit where the data does not actually exist.

Comment (MorpheusXAUT) [reply pkg/manager/impl/artifact_manager.go:314]: > Could not agree more! I noticed the `CreateArtifact` implementation can have this same issue (orphaned data on failure). I don't think this is common issue now, but if we find it is we will have to implement some kind of garbage collection mechanism.

Yeah, that could make sense - try to find orphaned data by comparing data store and database state. I'd like to keep that out of this PR/RFC implementation though just so we don't get too much feature creep.

> Yeah, If you have other thoughts I'm very open. This is the best approach that I could think of - I think a partial update is preferential to a cache hit where the data does not actually exist.

Definitely makes more sense than the current implementation. I'll see if I can come up with something else tomorrow, otherwise I'll change the order to your suggestion.

Comment (MorpheusXAUT) [reply pkg/manager/impl/artifact_manager.go:314]: @hamersaw Artifact data is now deleted after DB models have been created. I believe this should be enough for the main case we talked about (leaving "broken" artifact data entries without blob storage data).
Cleanup of potentially orphaned blob storage data is something for another PR, I'd say.

Comment (hamersaw) [reply pkg/manager/impl/artifact_manager.go:314]: I agree, thanks so much!

File: pkg/manager/impl/artifact_manager.go [hunk 2]
```diff
@@ -137,65 +144,83 @@ func (m *artifactManager) GetArtifact(ctx context.Context, request *datacatalog.
...
+}
+
+func (m *artifactManager) findArtifact(ctx context.Context, datasetID *datacatalog.DatasetID, queryHandle artifactQueryHandle) (models.Artifact, error) {
 	var artifactModel models.Artifact
-	switch request.QueryHandle.(type) {
-	case *datacatalog.GetArtifactRequest_ArtifactId:
-		logger.Debugf(ctx, "Get artifact by id %v", request.GetArtifactId())
-		artifactKey := transformers.ToArtifactKey(datasetID, request.GetArtifactId())
+
+	key := queryHandle.GetArtifactId()
```
Comment (hamersaw) [inline pkg/manager/impl/artifact_manager.go:183]: Is there any reason we changed this from using a switch over to `len(queryHandle.GetArtifcatId())` from `request.QueryHandle.(type)`. Was it required because of the addition of UpdateArtifactRequest handling?

Comment (MorpheusXAUT) [reply pkg/manager/impl/artifact_manager.go:183]: @hamersaw Yes, exactly. Unfortunately the type generated for the `UpdateArtifactRequest` is a `datacatalog.UpdateArtifactRequest_ArtifactId` (instead of a `datacatalog.GetArtifactRequest_ArtifactId`), however both implement `GetArtifactId() string`.
As the `oneof` interface is not exported in the generated protobuf definitions, we can't use a shared interface, so this was the easiest I came up with (which allows both to be passed as a single method parameter).
